AlternativeTextType as enum class with EnumTraits and serialization

angelaizg opened the pull request:
#### d84ac58bd76d15e40b36f215e7c8ecae110686d2
<pre>
AlternativeTextType as enum class with EnumTraits and serialization
<a href="https://bugs.webkit.org/show_bug.cgi?id=248174">https://bugs.webkit.org/show_bug.cgi?id=248174</a>
rdar://102309622

Making AlternativeTextType an enum class with EnumTraits and serialize it instead of casting to/from an int32_t. Prevents from assertion failure in void _NSWindowSetFrameIvar

Reviewed by Youenn Fablet

Canonical link: <a href="https://commits.webkit.org/257360@main">https://commits.webkit.org/257360@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ebad6030560c521efafc3e0361b384eae04b48a6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98287 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7495 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31425 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107737 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168007 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102230 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7999 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85015 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90862 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104323 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103941 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6025 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89584 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33101 "Passed tests") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/101833 "webkitpy-tests (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87895 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21001 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76045 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1454 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22521 "Found 1 new test failure: streams/readable-stream-default-controller-error.html (failure)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1398 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45136 "Found 30 new test failures: accessibility/aria-keyshortcuts.html, css3/supports-dom-api.html, fast/css/css-typed-om/style-property-map-set-CSSMathSum-value.html, fast/rendering/render-style-null-optgroup-crash.html, fast/text/text-edge-no-half-leading-simple.html, fast/text/text-edge-property-parsing.html, http/wpt/service-workers/fetch-service-worker-preload-changing-request.https.html, http/wpt/service-workers/fetch-service-worker-preload-download.https.html, http/wpt/service-workers/fetch-service-worker-preload.https.html, imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml.html ... (failure)") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5052 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6292 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41942 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2748 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on LayoutTests/ipc/WebPageProxy_CorrectionPanel-expected.txt:
```diff
@@ -0,0 +1,6 @@
+CONSOLE MESSAGE: ReferenceError: Can't find variable: IPC
```
AH... looks like this test is broken?

angelaizg replied:
It is supposed to happen, since the int sent is not in the enum cases

rniwa commented on Source/WebCore/loader/icon/IconLoader.cpp:
```diff
@@ -56,6 +56,7 @@ IconLoader::~IconLoader()
 
 void IconLoader::startLoading()
 {
+    WTFLogAlways("*************IconLoader::startLoading******************");
```
This seems like an unrelated change.

rniwa commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -336,7 +336,6 @@ ResourceErrorOr<CachedResourceHandle<CachedRawResource>> CachedResourceLoader::r
     default:
         ASSERT_NOT_REACHED();
     }
-
```
These whitespace changes seem untreated. Revert?

angelaizg replied:
Fixed

rniwa commented on Source/WebCore/platform/graphics/WebMResourceClient.cpp:
```diff
@@ -36,6 +36,7 @@ namespace WebCore {
 
 RefPtr<WebMResourceClient> WebMResourceClient::create(WebMResourceClientParent& parent, PlatformMediaResourceLoader& loader, ResourceRequest&& request)
 {
+    WTFLogAlways("*************WebMResourceClient::create********");
```
This seems like unrelated change.

youennf commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -9672,10 +9674,14 @@ void WebPageProxy::substitutionsPanelIsShowing(CompletionHandler<void(bool)>&& c
     // FIXME: Make AlternativeTextType an enum class with EnumTraits and serialize it instead of casting to/from an int32_t.
-    pageClient().showCorrectionPanel((AlternativeTextType)panelType, boundingBoxOfReplacedString, replacedString, replacementString, alternativeReplacementStrings);
+    if (panelType == AlternativeTextType::AlternativeTextTypeDictationAlternatives) {
+        ASSERT_IS_TESTING_IPC();
```
It would be good to add a test for this code path.

kkinnunen-apple replied:
Change this to 
`MESSAGE_CHECK(panelType != AlternativeTextType::AlternativeTextTypeDictationAlternatives);`

It's unclear to me if the test can be added to the layout test after you do this. You can check and see if it crashes or is skipped.

angelaizg replied:
Changes applied

youennf commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -858,11 +859,12 @@ void WebPageProxy::setResourceLoadClient(std::unique_ptr<API::ResourceLoadClient
     if (!m_injectedBundleClient)
         return;
-
+        
```
These two changes are not needed.

youennf commented on Source/WebKitLegacy/mac/WebCoreSupport/CorrectionPanel.mm:
```diff
@@ -35,18 +35,18 @@
     }
-    ASSERT_NOT_REACHED();
-    return NSCorrectionIndicatorTypeDefault;
+    
```
It seems we could probably share correctionIndicatorType with CorrectionPanel to reduce redundant code.

angelaizg replied:
Opened a new bug a new patch that would take care of this

youennf commented on LayoutTests/ipc/WebPageProxy_CorrectionPanel.html:
```diff
@@ -0,0 +1,4 @@
+<script>
```
This is probably not right, other IPC tests have:
`<!DOCTYPE html><!-- webkit-test-runner [ IPCTestingAPIEnabled=true ] -->`

This might explain the issue @rniwa spotted.

kkinnunen-apple replied:
If the test case is possible to be added, it could be named something like webpageproxy-correctionpanel-no-crash.html. See the other files in the directory for examples how to write the test.

kkinnunen-apple replied:
Also:
Originally the value was size of int32_t and the problem was that valid values were 0-3 but 655 was sent.

Now your code changes the value to size of uint8_t. 
So now the test just tests that the IPC wire format checks that the message is correct sized. The test will fail because int32_t is sent instead of uint8_t.

This means that the test case will not check the original underlying bug, e.g. sending values outside the allowed enum range.

So what you could do is to send couple of messages with different values > 3, but with uint8_t type. So this way you could ensure that the original bug is tested, e.g. that the function only allows the values in the enum.

Also, you could then add the case for the disallowed enum value, e.g.  `AlternativeTextTypeDictationAlternatives`.  This is with the caveat that it's not entirely clear to me if `MESSAGE_CHECK`ed values can be tested, or will that too crash.

youennf commented on LayoutTests/ipc/webpageproxy-correctionpanel-no-crash-expected.txt:
```diff
@@ -0,0 +1 @@
+
```
I would tend to do like the other tests and add something like
`Test passes if it does not crash.` in the html and which would show up there.

angelaizg replied:
I just added it in the parameters of that test

youennf commented on Source/WebKit/UIProcess/mac/CorrectionPanel.mm:
```diff
@@ -35,13 +35,13 @@
+    case WebCore::AlternativeTextType::AlternativeTextTypeSpellingSuggestions:
         return NSCorrectionIndicatorTypeGuesses;
-    case WebCore::AlternativeTextTypeDictationAlternatives:
+    case WebCore::AlternativeTextType::AlternativeTextTypeDictationAlternatives:
```
This one is now different from the WebKitLegacy one, let's keep them the same

youennf reviewed (commented):
Does it make sense to add a special test for AlternativeTextType::AlternativeTextTypeDictationAlternatives?

youennf commented on Source/WebCore/page/AlternativeTextClient.h:
```diff
@@ -38,11 +38,11 @@ enum ReasonForDismissingAlternativeText {
     AlternativeTextTypeReversion,
     AlternativeTextTypeSpellingSuggestions,
-    AlternativeTextTypeDictationAlternatives
+    DictationAlternatives
```
It seems a good idea to rename AlternativeTextTypeDictationAlternatives to DictationAlternatives.
Can we do the same for the others 3 as well now?

angelaizg replied:
Added changes

youennf replied:
Looks good now

youennf commented on Source/WebKit/UIProcess/mac/CorrectionPanel.mm:
```diff
@@ -35,18 +35,17 @@
         ASSERT_NOT_REACHED();
-        break;
+        return NSCorrectionIndicatorTypeDefault;
+            
```
Unneeded line

youennf commented on Source/WebKitLegacy/mac/WebCoreSupport/CorrectionPanel.mm:
```diff
@@ -35,18 +35,18 @@
         ASSERT_NOT_REACHED();
-        break;
+        return NSCorrectionIndicatorTypeDefault;
+
```
Unneeded line

youennf commented on Source/WebCore/page/AlternativeTextClient.h:
```diff
@@ -68,3 +68,17 @@ class AlternativeTextClient {
+        WebCore::AlternativeTextType::AlternativeTextTypeReversion,
+        WebCore::AlternativeTextType::AlternativeTextTypeSpellingSuggestions,
+        WebCore::AlternativeTextType::DictationAlternatives
+    
```
Unneeded line

youennf commented on LayoutTests/ipc/webpageproxy-correctionpanel-no-crash.html:
```diff
@@ -0,0 +1,14 @@
+    testRunner.dumpAsText();
+
+if (window.IPC)
+    IPC.sendMessage('UI',IPC.webPageProxyID,IPC.messages.WebPageProxy_ShowCorrectionPanel.name,[{type: 'int32_t',value: 120},{type: 'float',value: 134.3630587066791},{type: 'float',value: 659.5882936421311},{type: 'float',value: 249},{type: 'uint8_t',value: 3},{type: 'String',value: ""},{type: 'String',value: ""},{type: 'Vector',value: [[{type: 'String',value: "cancelOperation"}],[{type: 'String' ...
```
Shouldn't we pass a uint8_t whose value is 120 now that we are serialising a uint8_t enum?

angelaizg replied:
why is that?

youennf commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -9680,10 +9681,10 @@ void WebPageProxy::substitutionsPanelIsShowing(CompletionHandler<void(bool)>&& c
 {
     // FIXME: Make AlternativeTextType an enum class with EnumTraits and serialize it instead of casting to/from an int32_t.
-    pageClient().showCorrectionPanel((AlternativeTextType)panelType, boundingBoxOfReplacedString, replacedString, replacementString, alternativeReplacementStrings);
+    pageClient().showCorrectionPanel(panelType, boundingBoxOfReplacedString, replacedString, replacementString, alternativeReplacementStrings);
```
Please remove above FIXME.

angelaizg replied:
Added changes

youennf reviewed (approved):
r=me.
Please remove the no longer necessary FIXME.

angelaizg reviewed (commented):
Added the changes needed.

Commit:
AlternativeTextType as enum class with EnumTraits and serialization
https://bugs.webkit.org/show_bug.cgi?id=248174
rdar://102309622

Making AlternativeTextType an enum class with EnumTraits and serialize it instead of casting to/from an int32_t. Prevents from assertion failure in void _NSWindowSetFrameIvar

Reviewed by Youenn Fablet

Canonical link: https://commits.webkit.org/257360@main
